client/state/plugins/installed/selectors.js

```diff
@@ -66,65 +66,131 @@ export function requestPluginsError( state ) {
+export const getAllPluginsIndexedByPluginSlug = createSelector(
+	( state ) => {
+		if ( isRequestingForAllSites( state ) ) {
+			return {};
+		}
 
-			// We currently support fetching plugins per site and also fetching all plugins
-			// in bulk, aiming to optimize the UX in some flows.
-			if ( isRequestingForAllSites( state ) ) {
-				return memo;
+		return getSiteIdsThatHavePlugins( state ).reduce( ( plugins, siteId ) => {
+			if ( isRequesting( state, siteId ) ) {
+				return plugins;
 			}
 
-			const list = state.plugins.installed.plugins[ siteId ] || [];
-			list.forEach( ( item ) => {
-				const sitePluginInfo = pick( item, [ 'active', 'autoupdate', 'update', 'version' ] );
+			const pluginsForSite = state.plugins.installed.plugins[ siteId ] || [];
+			pluginsForSite.forEach( ( plugin ) => {
+				const sitePluginInfo = pick( plugin, [ 'active', 'autoupdate', 'update', 'version' ] );
 
-				memo[ item.slug ] = {
-					...memo[ item.slug ],
-					...item,
+				plugins[ plugin.slug ] = {
+					...plugins[ plugin.slug ],
+					...plugin,
 					sites: {
-						...memo[ item.slug ]?.sites,
+						...plugins[ plugin.slug ]?.sites,
 						[ siteId ]: sitePluginInfo,
 					},
 				};
 			} );
-			return memo;
-		},
-		{}
-	);
+			return plugins;
+		}, {} );
+	},
+	( state ) => [
+		isRequestingForAllSites( state ),
+		getSiteIdsThatHavePlugins( state ),
+		...getSiteIdsThatHavePlugins( state ).map( ( siteId ) => isRequesting( state, siteId ) ),
+	]
+);
 
-	if ( pluginFilter && _filters[ pluginFilter ] ) {
-		pluginList = filter( pluginList, _filters[ pluginFilter ] );
-	}
+/**
+ * The plugins here differ from the plugin objects found on state.plugins.installed.plugins in that each plugin
+ * object has data from all the different sites on it, whereas on state.plugins.installed.plugins they only have
+ * the state for the site which index they are under. The objects here are the same as those returned from
+ * getAllPluginsIndexedByPluginSlug, except they are indexed by siteId.
+ */
+export const getAllPluginsIndexedBySiteId = createSelector(
+	( state ) => {
+		const allPluginsIndexedByPluginSlug = getAllPluginsIndexedByPluginSlug( state );
+		const siteIdsThatHavePlugins = getSiteIdsThatHavePlugins( state );
+
+		const plugins = siteIdsThatHavePlugins.reduce( ( accumulator, siteId ) => {
+			accumulator[ siteId ] = {};
+			return accumulator;
+		}, {} );
+
+		return Object.values( allPluginsIndexedByPluginSlug ).reduce( ( accumulator, plugin ) => {
+			Object.keys( plugin.sites )
+				.map( ( siteId ) => Number( siteId ) )
+				.forEach( ( siteId ) => {
+					accumulator[ siteId ] = {
+						...accumulator[ siteId ],
+						[ plugin.slug ]: plugin,
+					};
+				} );
+
+			return accumulator;
+		}, plugins );
+	},
+	( state ) => [ getAllPluginsIndexedByPluginSlug( state ), getSiteIdsThatHavePlugins( state ) ]
+);
 
-	return sortBy( pluginList, ( item ) => item.slug.toLowerCase() );
-}
+export const getFilteredAndSortedPlugins = createSelector(
```

Comment:
`getPlugins` was a deceptively simple name compared to all the computation that was going on here and was arguably masking the very computationally expensive things being done previously, so I renamed it to `getFilteredAndSortedPlugins` to make it clear that there is more going on than simply getting plugins.